src/wallet/rpcwallet.cpp

```diff
@@ -4475,11 +4479,27 @@ static RPCHelpMan upgradewallet()
         version = request.params[0].get_int();
     }
     bilingual_str error;
-    if (!pwallet->UpgradeWallet(version, error)) {
-        throw JSONRPCError(RPC_WALLET_ERROR, error.original);
+    const int previous_version{pwallet->GetVersion()};
+    const bool wallet_upgraded{pwallet->UpgradeWallet(version, error)};
+    const int current_version{pwallet->GetVersion()};
+    std::string result;
+
+    if (wallet_upgraded) {
+        if (previous_version == current_version) {
+            result = "Already at latest version. Wallet version unchanged.";
+        } else {
+            result = strprintf("Wallet upgraded successfully from version %i to version %i.", previous_version, current_version);
+        }
     }
+
     UniValue obj(UniValue::VOBJ);
-    if (!error.empty()) {
+    obj.pushKV("wallet_name", pwallet->GetName());
+    obj.pushKV("previous_version", previous_version);
+    obj.pushKV("current_version", current_version);
+    if (!result.empty()) {
+        obj.pushKV("result", result);
+    } else {
+        CHECK_NONFATAL(!error.empty());
```

Comment:
style nit ca8cd893bb56bf5d455154b0498b1f58f77d20ed:

This should check (before the if) that there is either an error or a result, but not both.

```cpp

CHECK_NONFATAL(error.empty != result.empty());

Reply:
> style nit [ca8cd89](https://github.com/bitcoin/bitcoin/commit/ca8cd893bb56bf5d455154b0498b1f58f77d20ed):
> 
> This should check (before the if) that there is either an error or a result, but not both.

Good idea, that would be better.

Reply:
Done in e915a2a2c87dad5beb68a175b21a07136bcc21ca